# 14-2. wine dataset 실습

## 와인 분류 문제 : 와인 종류 분류하기

### 데이터 가져오기 및 살펴보기

#### load_wine import 해와서 wine 데이터를 로딩

In [6]:
from sklearn.datasets import load_wine
# sklearn 라이브러리의 datasets 패키지에서 load_wine 함수를 임포트함

wine = load_wine()
# 로드된 wine 데이터셋을 wine 변수에 저장

#### wine에는 어떤 정보들이 담겼을지, keys() 라는 메서드로 확인

In [7]:
wine.keys()
# wine 데이터셋에 담긴 정보 종류 확인

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

#### 데이터는 wine_data 변수에 저장한 후, 데이터의 크기를 확인

In [8]:
wine_data = wine.data
# keys에서 확인한 정보 중 data를 따로 wine_data 변수에 저장

print(wine_data.shape) 
# shape는 배열의 형상정보를 출력

(178, 13)


### 분석 : 
178개의 데이터가 각각 13개의 정보를 담고 있음

#### 샘플로 하나의 데이터만 확인

In [10]:
wine_data[0]
# 178개의 데이터 중 첫 번째 데이터 출력

array([1.423e+01, 1.710e+00, 2.430e+00, 1.560e+01, 1.270e+02, 2.800e+00,
       3.060e+00, 2.800e-01, 2.290e+00, 5.640e+00, 1.040e+00, 3.920e+00,
       1.065e+03])

#### 라벨(label) 또는 타겟(target) 확인 및 저장

In [11]:
wine_label = wine.target
# keys에서 확인한 정보 중 target을 따로 wine_label 변수에 저장

print(wine_label.shape)
wine_label

(178,)


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2])

### 분석:
길이와 형태를 확인하니 총 178개의 데이터가 들어있고, 각 값은 0, 1, 2로 나타나는 것을 확인 가능

#### 라벨의 이름을 target_names에서 확인

In [12]:
wine.target_names
# keys에서 확인한 정보 중 target_names를 변수에 따로 저장하지 않고 호출
# wine_label이 가진 0, 1, 2의 이름 확인

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

### 분석:
'class_0', 'class_1', 'class_2' 순서대로 담겨있음

이 순서 그대로 0이라면 class_0, 1이라면 class_1, 2라면 class_2를 나타냄

#### feature_names: 13개의 각 feature에 대한 설명

In [13]:
wine.feature_names
# keys에서 확인한 정보 중 feature_names를 변수에 따로 저장하지 않고 호출
# wine.data에서 확인한 30개 정보의 변수명 확인

['alcohol',
 'malic_acid',
 'ash',
 'alcalinity_of_ash',
 'magnesium',
 'total_phenols',
 'flavanoids',
 'nonflavanoid_phenols',
 'proanthocyanins',
 'color_intensity',
 'hue',
 'od280/od315_of_diluted_wines',
 'proline']

#### 데이터셋을 DataFrame으로 변환

DataFrame 을 만들면서 data에는 wine_data 넣어주고, 각 컬럼에는 feature_names로 이름을 붙여주기

In [14]:
import pandas as pd

wine_df = pd.DataFrame(data=wine_data, columns=wine.feature_names)
wine_df

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  
0                            3.92   1065.0  
1                            3.40   1050.0  
2                            3.17   1185.0  
3                            3.45   1480.0  
4                            2.93    735.0  
..                            ...      ...  
173                          1.74    740.0  
174                          1.56    750.0  
175                          1.56    835.0  
176                          1.62    840.0  
177                          1.60    560.0  

[178 rows x 13 columns]

#### label(정답) 컬럼을 추가

In [16]:
wine_df["label"] = wine.target
# wine_df에 label이라는 컬럼을 새로 추가
wine_df

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  label  
0                            3.92   1065.0      0  
1                            3.40   1050.0      0  
2                            3.17   1185.0      0  
3                            3.45   1480.0      0  
4                            2.93    735.0      0  
..                            ...      ...    ...  
173                          1.74    740.0      2  
174                          1.56    750.0      2  
175                          1.56    835.0      2  
176                          1.62    840.0      2  
177                          1.60    560.0      2  

[178 rows x 14 columns]

#### 결측치 확인

In [17]:
wine_df.isna().sum()

alcohol                         0
malic_acid                      0
ash                             0
alcalinity_of_ash               0
magnesium                       0
total_phenols                   0
flavanoids                      0
nonflavanoid_phenols            0
proanthocyanins                 0
color_intensity                 0
hue                             0
od280/od315_of_diluted_wines    0
proline                         0
label                           0
dtype: int64

### 분석:
데이터에서 결측치는 찾을 수 없다

#### 학습에 사용하는 training dataset과 모델의 성능을 평가하는 데 사용하는 test dataset으로 데이터셋을 나누기

In [19]:
from sklearn.model_selection import train_test_split
# sklearn model_selection패키지의 train_test_split 함수를 임포트

X_train, X_test, y_train, y_test = train_test_split(wine_data, 
                                                    wine_label, 
                                                    test_size=0.2, 
                                                    random_state=0)
# 나눠야 할 데이터(문제지, X): wine_data
# 데이터의 라벨(정답, y): wine_label
# wine_data, wine_label 각각 train:test = 8:2의 비율로 잘라서 
# X_train, X_test, y_train, y_test에 저장


X_train.shape, y_train.shape, X_test.shape, y_test.shape
# train의 형상정보 확인 # test의 형상정보 확인

((142, 13), (142,), (36, 13), (36,))

### 머신러닝 모델 적용

## DecisionTree 

In [20]:
from sklearn.tree import DecisionTreeClassifier # 싸이킷런에 있는 결정트리분류기를 사용하기 위해 불러오는 코드
from sklearn.metrics import classification_report #싸이킷런에 있는 분류 결과에 대한 시각화를 위해 쓰는 코드

decision_tree = DecisionTreeClassifier(random_state=1) #결정트리분류기의 객체를 만든다.
decision_tree.fit(X_train, y_train) # 분류기에 x와 y의 훈련 데이터를 넣어 훈련 시킨다.
y_pred = decision_tree.predict(X_test) # 훈련된 분류기에 X_test라는 테스트 데이터셋을 넣어 얼마나 예측했는지 확인한다.

decision_report = classification_report(y_test, y_pred) # 결과를 지표로 확인하기 위해 classification_report를 활용해 y_test, y_pred 값을 넣어 확인한다.
print(decision_report)

              precision    recall  f1-score   support

           0       0.93      1.00      0.97        14
           1       1.00      0.94      0.97        16
           2       1.00      1.00      1.00         6

    accuracy                           0.97        36
   macro avg       0.98      0.98      0.98        36
weighted avg       0.97      0.97      0.97        36



## RandomForest

In [21]:
from sklearn.ensemble import RandomForestClassifier #랜덤포레스트라는 분류기를 사용하기 위해 import

random_forest = RandomForestClassifier(random_state=2) # RandomForest분류기 객체를 생성
random_forest.fit(X_train, y_train) # 훈련
y_pred = random_forest.predict(X_test) # 예측


random_report = classification_report(y_test, y_pred)
print(random_report) # 결과 지표를 확인

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        14
           1       1.00      0.94      0.97        16
           2       0.86      1.00      0.92         6

    accuracy                           0.97        36
   macro avg       0.95      0.98      0.96        36
weighted avg       0.98      0.97      0.97        36



## Support Vector Machine (SVM)

In [22]:
from sklearn import svm #Support Vector Machine을 사용하기 위해 import
svm_model = svm.SVC() # 모델 객체를 만든다.

svm_model.fit(X_train, y_train) # 훈련
y_pred = svm_model.predict(X_test) # 예측

svm_report = classification_report(y_test, y_pred)
print(svm_report) # 결과 지표를 확인


              precision    recall  f1-score   support

           0       1.00      0.93      0.96        14
           1       0.80      0.75      0.77        16
           2       0.38      0.50      0.43         6

    accuracy                           0.78        36
   macro avg       0.72      0.73      0.72        36
weighted avg       0.81      0.78      0.79        36



## Stochastic Gradient Descent Classifier (SGDClassifier)

In [23]:
from sklearn.linear_model import SGDClassifier #선형분류기인 SGDClassifier를 사용하기 위한 import
sgd_model = SGDClassifier() # 모델 객체 생성

sgd_model.fit(X_train, y_train) # sgd모델로 훈련데이터로 훈련시킨다.
y_pred = sgd_model.predict(X_test)# 그 모델로 test데이터를 사용해 예측

sgd_report = classification_report(y_test, y_pred)
print(sgd_report) # 결과 지표를 확인

              precision    recall  f1-score   support

           0       0.58      1.00      0.74        14
           1       0.82      0.56      0.67        16
           2       1.00      0.17      0.29         6

    accuracy                           0.67        36
   macro avg       0.80      0.58      0.56        36
weighted avg       0.76      0.67      0.63        36



## Logistic Regression

In [24]:
from sklearn.linear_model import LogisticRegression # 선형분류기인 LogisticRegression를 사용하기 위한 import
logistic_model = LogisticRegression() # 모델 객체 생성

logistic_model.fit(X_train, y_train) #LogisticRegression모델로 훈련데이터를 가지고 훈련시킨다.
y_pred = logistic_model.predict(X_test) # 예측

logistic_report = classification_report(y_test, y_pred)
print(logistic_report) # 결과 지표를 확인

              precision    recall  f1-score   support

           0       0.93      0.93      0.93        14
           1       0.93      0.88      0.90        16
           2       0.86      1.00      0.92         6

    accuracy                           0.92        36
   macro avg       0.91      0.93      0.92        36
weighted avg       0.92      0.92      0.92        36



/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [25]:
#### 다양한 모델들의 예측값의 결가지표 비교하기
print('< Decision Tree >')
print(decision_report)

print('< Random Forest >')
print(random_report)

print('< SVM >')
print(svm_report)

print('< SGD >')
print(sgd_report)

print('< Logistic Regression >')
print(logistic_report)

< Decision Tree >
              precision    recall  f1-score   support

           0       0.93      1.00      0.97        14
           1       1.00      0.94      0.97        16
           2       1.00      1.00      1.00         6

    accuracy                           0.97        36
   macro avg       0.98      0.98      0.98        36
weighted avg       0.97      0.97      0.97        36

< Random Forest >
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        14
           1       1.00      0.94      0.97        16
           2       0.86      1.00      0.92         6

    accuracy                           0.97        36
   macro avg       0.95      0.98      0.96        36
weighted avg       0.98      0.97      0.97        36

< SVM >
              precision    recall  f1-score   support

           0       1.00      0.93      0.96        14
           1       0.80      0.75      0.77        16
           2       0.38      0.

### 분석:

실습의 목표는 와인의 종류를 분류하여 찾아내는 것이다.

조금 더 세부적인 상황으로 와인의 특징을 기입하고 해당 와인이 A종류의 와인인지 확인하여 출고하는 경우라고 가정하자.

이 경우 실제 A종류의 와인인데 아니라고 잘못된 판단을 하는 것보다 B종류의 와인을 A종류의 와인이라고 하는 잘못 판단하는 경우가 더 치명적인 상황이 될 것이다. 따라서 오류라도 현재 가정된 상황에서는 positive를 잡아내는 데에 실패하는 오류보다 negative를 잡아내는 데에 실패하는 오류의 치명도와 중요도가 훨씬 크다. 

현재 가정된 상황에서는 실제 negative인데 positive로 잘못 판단하는 경우가 가장 치명적인 오류이고, 이 경우가 가장 적어야한다. 따라서 FP이 낮을수록 수치가 커지는 Precision 값이 가장 중요한 척도라고 할 수 있다.

   
위의 각 모델 예측 결과의 confusion matrix를 살펴보면,

    Decision Tree의 경우 각 와인의 종류별로 precision 값이 0.93, 1.00, 1.00이고 f1-score는 0.97, 0.97, 1.00이다.
    
    Random Forest의 경우 각 와인의 종류별로 precision 값이 1.00, 1.00, 0.86, 이고 f1-score는 1.00, 0.97, 0.92이다.

    SVM의 경우 각 와인의 종류별로 precision 값이 1.00, 0.80, 0.38, 이고 f1-score는 0.96, 0.77, 0.43이다.

    SGD의 경우 각 와인의 종류별로 precision 값이 0.58, 0.82, 1.00, 이고 f1-score는 0.74, 0.67, 0.29이다.

    Logistic Regression의 경우 각 와인의 종류별로 precision 값이 0.93, 0.93, 0.86, 이고 f1-score는 0.93, 0.90, 0.92이다.

#### 즉, 정확도도 높으며 거짓 양성 판정의 확률도 가장 적은 Decision Tree모델이 가장 적절한 예측 모델이라고 할 수 있다. 
